File: src/tests/java-core/net/sf/xmlunit/builder/InputTest.java
/*
  This file is licensed to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
package net.sf.xmlunit.builder;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileReader;
import java.net.URI;
import java.net.URL;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.transform.Source;
import org.w3c.dom.Document;
import net.sf.xmlunit.TestResources;
import net.sf.xmlunit.util.Convert;
import org.junit.Test;

import static org.hamcrest.core.Is.*;
import static org.hamcrest.core.IsInstanceOf.*;
import static org.hamcrest.core.IsNull.*;
import static org.junit.Assert.*;

public class InputTest {

    private static Document parse(Source s) throws Exception {
        DocumentBuilder b =
            DocumentBuilderFactory.newInstance().newDocumentBuilder();
        return b.parse(Convert.toInputSource(s));
    }

    @Test public void shouldParseADocument() throws Exception {
        Document d = parse(Input.fromFile(TestResources.ANIMAL_FILE).build());
        Source s = Input.fromDocument(d).build();
        allIsWellFor(s);
    }

    @Test public void shouldParseAnExistingFileByName() throws Exception {
        Source s = Input.fromFile(TestResources.ANIMAL_FILE).build();
        allIsWellFor(s);
        assertEquals(toFileUri(TestResources.ANIMAL_FILE), s.getSystemId());
    }

    @Test public void shouldParseAnExistingFileByFile() throws Exception {
        Source s = Input.fromFile(new File(TestResources.ANIMAL_FILE)).build();
        allIsWellFor(s);
        assertEquals(toFileUri(TestResources.ANIMAL_FILE), s.getSystemId());
    }

    @Test public void shouldParseAnExistingFileFromStream() throws Exception {
        FileInputStream is = null;
        try {
            is = new FileInputStream(TestResources.ANIMAL_FILE);
            allIsWellFor(Input.fromStream(is).build());
        } finally {
            if (is != null) {
                is.close();
            }
        }
    }

    @Test public void shouldParseAnExistingFileFromReader() throws Exception {
        FileReader r = null;
        try {
            r = new FileReader(TestResources.ANIMAL_FILE);
            allIsWellFor(Input.fromReader(r).build());
        } finally {
            if (r != null) {
                r.close();
            }
        }
    }

    @Test public void shouldParseString() throws Exception {
        allIsWellFor(Input.fromMemory(new String(readTestFile(), "UTF-8"))
                     .build());
    }

    @Test public void shouldParseBytes() throws Exception {
        allIsWellFor(Input.fromMemory(readTestFile()).build());
    }

    @Test public void shouldParseFileFromURIString() throws Exception {
        allIsWellFor(Input.fromURI("file:" + TestResources.ANIMAL_FILE).build());
    }

    @Test public void shouldParseFileFromURI() throws Exception {
        allIsWellFor(Input.fromURI(new URI("file:" + TestResources.ANIMAL_FILE))
                     .build());
    }

    @Test public void shouldParseFileFromURL() throws Exception {
        allIsWellFor(Input.fromURL(new URL("file:" + TestResources.ANIMAL_FILE))
                     .build());
    }

    @Test public void shouldParseATransformationFromSource() throws Exception {
        Source input = Input.fromMemory("<animal>furry</animal>").build();
        Source s = Input.byTransforming(input)
            .withStylesheet(Input.fromFile("src/tests/resources/animal.xsl")
                            .build())
            .build();
        allIsWellFor(s, "furry");
    }

    @Test public void shouldParseATransformationFromBuilder() throws Exception {
        Input.Builder input = Input.fromMemory("<animal>furry</animal>");
        Source s = Input.byTransforming(input)
            .withStylesheet(Input.fromFile("src/tests/resources/animal.xsl"))
            .build();
        allIsWellFor(s, "furry");
    }

    private static void allIsWellFor(Source s) throws Exception {
        allIsWellFor(s, "animal");
    }

    private static void allIsWellFor(Source s, String rootElementName)
        throws Exception {
        assertThat(s, notNullValue());
        Document d = parse(s);
        assertThat(d, notNullValue());
        assertThat(d.getDocumentElement().getTagName(), is(rootElementName));
    }

    private static byte[] readTestFile() throws Exception {
        FileInputStream is = new FileInputStream(TestResources.ANIMAL_FILE);
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        byte[] buffer = new byte[1024];
        int read = -1;
        while ((read = is.read(buffer)) >= 0) {
            if (read > 0) {
                bos.write(buffer, 0, read);
            }
        }
        is.close();
        return bos.toByteArray();
    }

    private static String toFileUri(String fileName) {
        return new File(fileName).toURI().toString();
    }
}

File: src/tests/net-core/builder/InputTest.cs
/*
  This file is licensed to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
using System;
using System.IO;
using System.Text;
using System.Xml;
using NUnit.Framework;

namespace net.sf.xmlunit.builder {

    [TestFixture]
    public class InputTest {

        private const string TEST_FILE = "../../../src/tests/resources/test1.xml";

        private static XmlDocument Parse(ISource s) {
            XmlDocument d = new XmlDocument();
            d.Load(s.Reader);
            return d;
        }

        [Test] public void ShouldParseADocument() {
            XmlDocument d = Parse(Input.FromFile(TEST_FILE).Build());
            AllIsWellFor(Input.FromDocument(d).Build());
        }

        [Test] public void ShouldParseAnExistingFileByName() {
            AllIsWellFor(Input.FromFile(TEST_FILE).Build());
        }

        [Test] public void ShouldParseAnExistingFileFromStream() {
            using (FileStream fs = new FileStream(TEST_FILE, FileMode.Open,
                                                  FileAccess.Read)) {
                AllIsWellFor(Input.FromStream(fs).Build());
            }
        }

        [Test] public void ShouldParseAnExistingFileFromReader() {
            using (StreamReader r = new StreamReader(TEST_FILE)) {
                AllIsWellFor(Input.FromReader(r).Build());
            }
        }

        [Test] public void ShouldParseString() {
            AllIsWellFor(Input.FromMemory(Encoding.UTF8.GetString(ReadTestFile()))
                         .Build());
        }

        [Test] public void ShouldParseBytes() {
            AllIsWellFor(Input.FromMemory(ReadTestFile()).Build());
        }

        [Ignore("looks as if file-URIs didn't work, revisit")]
        [Test] public void ShouldParseFileFromURIString() {
            AllIsWellFor(Input.FromURI("file:" + TEST_FILE).Build());
        }

        [Ignore("looks as if file-URIs didn't work, revisit")]
        [Test] public void ShouldParseFileFromURI() {
            AllIsWellFor(Input.FromURI(new Uri("file:" + TEST_FILE)).Build());
        }

        [Test] public void ShouldParseATransformationFromSource() {
            ISource input = Input.FromMemory("<animal>furry</animal>").Build();
            ISource s = Input.ByTransforming(input)
                .WithStylesheet(Input.FromFile("../../../src/tests/resources/animal.xsl")
                                .Build())
                .Build();
            Assert.That(s, Is.Not.Null);
            XmlDocument d = Parse(s);
            Assert.That(d, Is.Not.Null);
            Assert.That(d.DocumentElement.Name, Is.EqualTo("furry"));
        }

        [Test] public void ShouldParseATransformationFromBuilder() {
            Input.IBuilder input = Input.FromMemory("<animal>furry</animal>");
            ISource s = Input.ByTransforming(input)
                .WithStylesheet(Input.FromFile("../../../src/tests/resources/animal.xsl"))
                .Build();
            Assert.That(s, Is.Not.Null);
            XmlDocument d = Parse(s);
            Assert.That(d, Is.Not.Null);
            Assert.That(d.DocumentElement.Name, Is.EqualTo("furry"));
        }

        private static void AllIsWellFor(ISource s) {
            Assert.That(s, Is.Not.Null);
            XmlDocument d = Parse(s);
            Assert.That(d, Is.Not.Null);
            Assert.That(d.DocumentElement.Name, Is.EqualTo("animal"));
        }

        private static byte[] ReadTestFile() {
            using (FileStream fs = new FileStream(TEST_FILE, FileMode.Open,
                                                  FileAccess.Read))
            using (MemoryStream ms = new MemoryStream()) {
                byte[] buffer = new byte[1024];
                int read = -1;
                while ((read = fs.Read(buffer, 0, buffer.Length)) > 0) {
                    ms.Write(buffer, 0, read);
                }
                return ms.ToArray();
            }
        }
    }
}